source/common/http/async_client_impl.h

```diff
@@ -466,25 +498,68 @@ class AsyncStreamImpl : public AsyncClient::Stream,
   friend class AsyncClientImplUnitTest;
 };
 
-class AsyncRequestImpl final : public AsyncClient::Request,
-                               AsyncStreamImpl,
-                               AsyncClient::StreamCallbacks {
+class AsyncRequestSharedImpl : public virtual AsyncClient::Request,
```

Comment:
sorry is this just splitting out a base class and having one subclass for old and one for new?
I assume that just doing the streaming vs not inline was uglier?
If we're going this route we should definitely comment what the structure is and add TODOs for clean up after runtime guard is gone.

Reply:
Right so now there's two separate methods in AsyncClient: `send` which expects a full request, and returns something that can only be cancelled (AsyncClient::Request*), and `startRequest` which expects headers only, and returns a whole stream that can be interacted with (AsyncClient::OngoingRequest*). The implementations of these two have a lot in common, hence the shared base with a protected constructor.